data-prepper-plugins/key-value-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -179,79 +230,219 @@ private void validateKeySets(final Set<String> includeSet, final Set<String> exc
+                } catch (Exception e) {
+                    LOG.error("Recursive parsing ran into an unexpected error, treating message as non-recursive");
                 }
+            } else {
+                outputMap.putAll(createNonRecursedMap(groups));
+            }
+
+            executeConfigs(outputMap, parsedMap);
 
-                if (excludeKeysSet.contains(key)) {
-                    LOG.debug("Key is being excluded: '{}'", key);
+            for (Map.Entry<String,Object> pair : defaultValuesMap.entrySet()) {
+                if (parsedMap.containsKey(pair.getKey())) {
+                    LOG.debug("Skipping already included default key: '{}'", pair.getKey());
                     continue;
                 }
+                parsedMap.put(pair.getKey(), pair.getValue());
+            }
 
-                if(keyValueProcessorConfig.getDeleteKeyRegex() != null && !Objects.equals(keyValueProcessorConfig.getDeleteKeyRegex(), "")) {
-                    key = key.replaceAll(keyValueProcessorConfig.getDeleteKeyRegex(), "");
-                }
-                key = keyValueProcessorConfig.getPrefix() + key;
+            recordEvent.put(keyValueProcessorConfig.getDestination(), parsedMap);
+        }
 
-                if (terms.length == 2) {
-                    value = terms[1];
-                } else {
-                    LOG.debug("Unsuccessful match: '{}'", terms[0]);
-                    value = keyValueProcessorConfig.getNonMatchValue();
-                }
+        return records;
+    }
+
+    private ObjectNode recurse(String input, ObjectMapper mapper) {
+        Stack<Character> bracketStack = new Stack<Character>();
+        Map<Character, Character> bracketMap = initBracketMap();
+        int pairStart = 0;
+
+        ArrayList<String> pairs = new ArrayList<String>();
+        ObjectNode root = mapper.createObjectNode();
+
+        for (int i = 0; i < input.length(); i++) {
+            if (bracketMap.containsKey(input.charAt(i))) {
+                bracketStack.push(input.charAt(i));
+            }
 
-                if(value != null
-                        && value instanceof String
-                        && keyValueProcessorConfig.getDeleteValueRegex() != null
-                        && !Objects.equals(keyValueProcessorConfig.getDeleteValueRegex(), "")) {
-                    value = ((String)value).replaceAll(keyValueProcessorConfig.getDeleteValueRegex(), "");
+            if (bracketMap.containsValue(input.charAt(i)) && !bracketStack.isEmpty()) {
+                if (bracketMap.get(bracketStack.peek()) == input.charAt(i)) {
+                    bracketStack.pop();
                 }
+            }
+
+            if (bracketStack.isEmpty() && input.charAt(i) == fieldDelimiterPattern.toString().charAt(0)) {
+                String pair = input.substring(pairStart, i);
+                pairs.add(pair);
+                pairStart = i + fieldDelimiterPattern.toString().length();
```

Comment:
I was thinking since there's the chance that the field delimiter can be set to a length of characters (eg. "==" or something similar), we want to increment by the length of the pattern to be safe